### DISCUSSION TITLE
Using cogs with `app_commands.command()`

### DISCUSSION THREAD

**Discussion by hitblast** [Rapptz/discord.py#8372] [Q&A] [answered]
What is the proper way to create application commands inside a cog (`commands.Cog`)? I'm new to the syntax of discord.py 2.0.0 so I'm looking for some possible answers to this silly question :P

**Comment by LostLuma** [comment, marked as answer]
You can create commands inside cogs by simply defining them inside and adding the `self` argument, similarly to how it works in the commands extension:
```py
@discord.app_commands.command()
async def upload(interaction: discord.Interaction, file: discord.Attachment) -> None:
    ...

class Example(commands.Cog):
    @discord.app_commands.command()
    async def upload(interaction: discord.Interaction, file: discord.Attachment) -> None:
        ....
```
When the cog is loaded it will be added to the [bot's command tree](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html#discord.ext.commands.Bot.tree), and you will be able to use and sync it just like all other app commands.

**Reply by hitblast** [reply]
Seems just about it. What would be the case if I want to invoke something *before the command runs*?  When it came to `commands.Context`, I used to do something similar to this:

```python
async def cog_before_invoke(ctx: commands.Context) -> None:
    ...
```

**Reply by LostLuma** [reply]
There's no built-in way to do this currently. You can however pretty easily create a decorator that wraps the command callback and apply it to every command that needs it.

**Reply by hitblast** [reply]
> There's no built-in way to do this currently. You can however pretty easily create a decorator that wraps the command callback and apply it to every command that needs it.

I see. Marking this conversation as answered, meanwhile could you please show me an example of wrapping the callback? I'm a little overwhelmed as of now :D

**Reply by LostLuma** [reply]
I've made a simple decorator that always responds with `h` to any command decorated with it as an example.
You can however do anything you need here, the cog is passed as `self` so it should be as useful as the callback you used to use.

I'm unsure where I can find a good resource on learning decorators currently, but they're definitively worth looking into!

```py
import functools

def send_h_before(f):
  @functools.wraps(f)
  async def callback(self, interaction: discord.Interaction, *args, **kwargs) -> None:
    await interaction.response.send_message('h')
    await f(self, interaction, *args, **kwargs)

  return callback

class Example(commands.Cog):
  @discord.app_commands.command()
  @send_h_before  # Make sure this is the last decorator in the list
  async def upload(self, interaction: discord.Interaction, attachment: discord.Attachment) -> None:
    ...
```

**Reply by hitblast** [reply]
Thanks!

**Reply by hitblast** [reply]
@LostLuma A couple last things I would like to ask, 

- Is there any way to develop a bot *only for application commands*? 
- What would be the case for cogs when we want to handle errors with them?

**Reply by LostLuma** [reply]
`Is there any way to develop a bot only for application commands?`
I'm not sure I understand the question, no one is forcing you to add text commands to your bot?
Additionally, you can pass `help_command=None` to the bot constructor to also remove the default help command.

`What would be the case for cogs when we want to handle errors with them?`
There's a new [`cog_app_command_error`](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html#discord.ext.commands.Cog.cog_app_command_error) method you can add to your cogs to handle app command errors.

**Reply by hitblast** [reply]
> `Is there any way to develop a bot only for application commands?` I'm not sure I understand the question, no one is forcing you to add text commands to your bot? Additionally, you can pass `help_command=None` to the bot constructor to also remove the default help command.
> 
> `What would be the case for cogs when we want to handle errors with them?` There's a new [`cog_app_command_error`](https://discordpy.readthedocs.io/en/stable/ext/commands/api.html#discord.ext.commands.Cog.cog_app_command_error) method you can add to your cogs to handle app command errors.

Well speaking of the first question, either of these two classes: `commands.Bot` and `commands.AutoShardedBot` are requiring me to pass a prefix, on the other hand it's not when I'm using `discord.Client`.

**Reply by Soheab** [reply]
You can pass an empty iterable / `None` to the `command_prefix=` keyword argument, but that won't disable the command parsing etc.

**Reply by atraboulsi27** [reply]
@LostLuma I know this is an old issue but was there an official implementation for this as of today? I feel it is a very basic feature to have especially if such a thing is already implemented through `before_invoke` with commands not part of a Cog. It would be also helpful to have the ability to wrap the Cog itself rather than repeating the decorator inside its functions (in cases where I need to have a global middeware)

Your solution worked perfectly though thanks! :)

**Reply by LostLuma** [reply]
Not as far as I'm aware, but I've also not kept up with discord.py development as much recently.

**Reply by atraboulsi27** [reply]
Deal, thanks
